referenceframe/urdf.go

```diff
@@ -0,0 +1,299 @@
+				XMLName xml.Name `xml:"box"`
+				Size    string   `xml:"size,attr"` // "x y z" format, in meters
+			} `xml:"box"`
+			Sphere struct {
+				XMLName xml.Name `xml:"sphere"`
+				Radius  float64  `xml:"radius,attr"` // in meters
+			} `xml:"sphere"`
+		} `xml:"geometry"`
+	} `xml:"collision"`
+}
+
+type UrdfJoint struct {
+	XMLName xml.Name `xml:"joint"`
+	Name    string   `xml:"name,attr"`
+	Type    string   `xml:"type,attr"`
+	Origin  struct {
+		XMLName xml.Name `xml:"origin"`
+		RPY     string   `xml:"rpy,attr"` // Fixed frame angle "r p y" format, in radians
+		XYZ     string   `xml:"xyz,attr"` // "x y z" format, in meters
+	} `xml:"origin"`
+	Parent struct {
+		XMLName xml.Name `xml:"parent"`
+		Link    string   `xml:"link,attr"`
+	} `xml:"parent"`
+	Child struct {
+		XMLName xml.Name `xml:"child"`
+		Link    string   `xml:"link,attr"`
+	} `xml:"child"`
+	Axis struct {
+		XMLName xml.Name `xml:"axis"`
+		XYZ     string   `xml:"xyz,attr"` // "x y z" format, in meters
+	} `xml:"axis"`
+	Limit struct {
+		XMLName xml.Name `xml:"limit"`
+		Lower   float64  `xml:"lower,attr"` // translation limits are in meters, revolute limits are in radians
+		Upper   float64  `xml:"upper,attr"` // translation limits are in meters, revolute limits are in radians
+	} `xml:"limit"`
+}
+
+// ParseURDFFile will read a given file and parse the contained URDF XML data into an equivalent ModelConfig struct.
+func ParseURDFFile(filename, modelName string) (Model, error) {
+	//nolint:gosec
+	xmlData, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to read URDF file")
+	}
+	return ConvertURDFToConfig(xmlData, modelName)
+}
+
+// ConvertURDFToConfig will transfer the given URDF XML data into an equivalent ModelConfig. Direct unmarshaling in the
+// same fashion as ModelJSON is not possible, as URDF data will need to be evaluated to accommodate differences
+// between the two kinematics encoding schemes.
+func ConvertURDFToConfig(xmlData []byte, modelName string) (Model, error) {
+	// empty data probably means that the read URDF has no actionable information
+	if len(xmlData) == 0 {
+		return nil, ErrNoModelInformation
+	}
+
+	mc := &ModelConfig{}
+	urdf := &URDFConfig{}
+	err := xml.Unmarshal(xmlData, urdf)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to convert URDF data to equivalent URDFConfig struct")
+	}
+
+	if modelName == "" {
+		modelName = urdf.Name
+	}
+
+	parentMap := map[string]string{}
+
+	// Migrate URDF elements into an equivalent ModelConfig representation
+	mc.Name = modelName
+	mc.KinParamType = "SVA"
+
+	// Handle joints
+	for _, jointElem := range urdf.Joints {
+		// Checking for reserved names in this or adjacent elements
+		if jointElem.Name == World {
+			return nil, errors.New("Joints with the name 'world' are not supported by config parsers")
+		}
+
+		// Relationship tracking
+		parentMap[jointElem.Name] = jointElem.Parent.Link
+		parentMap[jointElem.Child.Link] = jointElem.Name
```

Comment:
Won't this double-assign each frame? Once when we see a frame, and once when we see its parent?

If so, we should do a check whether each of these are in the map already, and if so, we should verify that their intended value matches what it already is.

Reply:
Since these relationships are assigned when striding the URDFs joints (and, by convention, there must be a link in between each joint), you end up assigning the adjacent links exactly twice*: Once as a child and once as a parent.

parent[joint1] -> link1 (first assignment)
parent[link2] -> link2 (second assignment)

*Except for the most proximal and most distal links, and there are attempts to catch the unique cases which may emerge as part of ignoring **world** links, etc.

Reply:
I'm not sure I'm following; is your example incorrect with link2?

Reply:
Yes, sorry that has been corrected in the comment.

Reply:
Ah; so each item is assigned to be a key or value only once, because we're iterating through the joints here only.

Can you confirm we will never need to add links in the same way? There can never be a link whose parent is another link?